agent/util-scripts/pbench-tool-meister-start

```diff
@@ -244,13 +226,530 @@ def waitpid(pid):
+                    host,
+                    pid,
+                    exit_status,
+                )
+            else:
+                successes += 1
+
+    assert tm_count == len(tool_group.hostnames) and tm_count == (
+        successes + failures
+    ), f"Logic bomb!  Number of successes ({successes}) and failures ({failures}) for TM creation don't add up (should be {tm_count})"
+
+    if failures > 0:
+        raise StartTmsErr(
+            "failures encountered creating tool miesters", ReturnCode.TMFAILURES
+        )
+    if successes != tm_count:
+        raise StartTmsErr(
+            f"number of created Tool Meisters, {successes}, does not"
+            f" match the expected number of Tool Meisters, {tm_count}",
+            ReturnCode.TMMISSING,
+        )
+
+
+class BaseServer:
+    """BaseServer - abstract base class for common code shared between the
+    ToolDataSink and RedisServer classes.
+    """
+
+    def_port = None
+    bad_port_code = None
+    bad_host_code = None
+    name = None
+
+    class Err(ReturnCode.Err):
+        """BaseServer.Err - derived from ReturnCode.Err, specifically raised by
+        BaseServer and its derived classes.
+        """
+
+        pass
+
+    def __init__(self, spec: str, def_host_name: str):
+        """__init__ - from the given IP/port specification given, determine the
+        IP:port for binding (listening) and the IP:port for connecting.
+
+        The IP/port specification can be given in one of two forms:
+
+          - `<ip>:<port>'
+            * where the same ip address and port are used for binding and
+              connecting
+          - `<bind ip>:<port>;<connect ip>:<port>`
+            * where a semi-colon separates the bind ip/port from the connecting
+              ip/port
+
+        In either case, a missing port (bare colon, optional) indicates the
+        default port should be used. If no IP address is given, the default
+        host name is used.
+
+        No attempt is made to verify that the IP address resolves, or that it
+        is reachable, though we do check they are syntactically valid.
+        """
+        _spec = spec if spec else def_host_name
+        parts = _spec.split(";", 1)
+        pairs = []
+        for part in parts:
+            host_port_parts = part.rsplit(":", 1)
+            if len(host_port_parts) == 1:
+                port = self.def_port
+            else:
+                try:
+                    port = int(host_port_parts[1])
+                except ValueError as exc:
+                    if host_port_parts[1] == "":
+                        port = self.def_port
+                    else:
+                        raise self.Err(
+                            f"Bad port specified for {self.name} in '{spec}'",
+                            self.bad_port_ret_code,
+                        ) from exc
+            host = host_port_parts[0] if host_port_parts[0] else def_host_name
+            if validate_hostname(host) != 0:
```

Comment:
If `host` contains an IPv6 address, it will (presumably) be surrounded by square brackets which need to be removed prior to validating it.